src/mysql.erl

```diff
@@ -175,32 +181,86 @@ start_link(Options) ->
          Query :: iodata(),
          Result :: query_result().
 query(Conn, Query) ->
-    query_call(Conn, {query, Query}).
+    query_call(Conn, {query, Query, undefined, undefined}).
 
 %% @doc Depending on the 3rd argument this function does different things.
 %%
-%% If the 3rd argument is a list, it executes a parameterized query. This is
-%% equivallent to query/4 with the query timeout as given to start_link/1.
+%% If the 3rd argument is a list, it executes a parameterized query and applies
+%% no filtering/mapping of the result rows. This is equivallent to query/5
+%% with undefined as filter/map function and the query timeout as given to
+%% start_link/1.
+%%
+%% If the 3rd argumeent is a function, it is used to filter/map the result
+%% rows. This is equivalent to query/4 with the query timeout given to
+%% start_link/1.
 %%
 %% If the 3rd argument is a timeout, it executes a plain query with this
-%% timeout.
+%% timeout and applies no filtering/mapping of the resultset rows.
 %%
 %% The return value is the same as for query/2.
 %%
 %% @see query/2.
 %% @see query/4.
--spec query(Conn, Query, Params | Timeout) -> Result
+-spec query(Conn, Query, Params | FilterMap | Timeout) -> Result
     when Conn :: connection(),
          Query :: iodata(),
          Timeout :: timeout(),
          Params :: [term()],
+         FilterMap :: query_filtermap(),
          Result :: query_result().
 query(Conn, Query, Params) when is_list(Params) ->
-    query_call(Conn, {param_query, Query, Params});
+    query_call(Conn, {param_query, Query, Params, undefined, undefined});
+query(Conn, Query, FilterMap) when FilterMap=:=undefined;
```

Comment:
I think I slightly disagree here, or at least I'm uncertain when it comes to conditionally using a function or not (the exmaples below should make clear what I'm trying to say).

If we don't allow `undefined` (or whatever else to indicate that no function is to be used, I'll stick to `undefined` for the time being), it has to be done either like this:
```erlang
Res=
case X of
    foo -> mysql:query(Conn, Query, fun ([A|_]) -> A<10 end);
    bar -> mysql:query(Conn, Query)
end
```
or like this:
```erlang
FilterMap=
case X of
    foo -> fun ([A|_]) -> A<10 end;
    bar -> fun (_) -> true end
end,
Res=mysql:query(Conn, Query, FilterMap)
```

If we allow `undefined`, this is also possible, on top of the other two approaches:
```erlang
FilterMap=
case X of
    foo -> fun ([A|_]) -> A<10 end;
    bar -> undefined
end,
Res=mysql:query(Conn, Query, FilterMap)
```

Reply:
On second thought, this would be an inconsistency with the current timeout behavior, since it is not possible to specify `default_timeout` or an equivalent... Therefore, I'll disallow `undefined` for the `FilterMap` parameter as you suggested.

Reply:
I don't see that there could be a problem. Before, it would allow an `integer` or `infinity` for timeouts, or a list for params. Now it will additionally allow `default_timeout` for timeouts, `no_params` for params, and `no_filtermap_fun` or a one-ary or two-ary function for filtermap. All things are additional, nothing has been removed.